snowfakery_samples/npsp/RD_npsp.recipe.yml

```diff
@@ -1,72 +1,123 @@
-- object: Account
-  nickname: A Company
-  fields:
-    Name:
-      fake: last_name
-    RecordType: Organization
+### ------------- [          ORDER OF INSERT         ]------------ ###
+# 1-0. Org Account (5)
+# 1-1. Contact (5)
+# 1-2. Define MACRO for npe03__Recurring_Donation__c
+# 1-3. npe03__Recurring_Donation__c (5) - Org Account
+# 1-4. npe03__Recurring_Donation__c (5) - Household
 
+### ----------- [ RD_npsp.recipe.yml Summary ]---------- ###
+# This recipe creates a set of 10 Recurring Donation records, at random # of Open-ended and Fixed-length donations, 
+# related to 5 Contact and 5 Organization Account.
+# This recipe is specifically for Enhanced Recurring Donations
 
-- object: Contact
-  fields:
-    FirstName:
-      fake: first_name
-    LastName:
-      fake: last_name
+### ----------- [               CLI RUN                ]---------- ###
+
+# cci task run generate_and_load_from_yaml --generator_yaml snowfakery_samples/npsp/RD_npsp.recipe.yml --num_records 300 --num_records_tablename npe03__Recurring_Donation__c --org qa
+# snowfakery --output-format json --output-file src/foo.json snowfakery_samples/npsp/RD_npsp.recipe.yml
+
+### ----------- [            Recipe Starts             ]---------- ###
 
 - macro: commomfields
   fields:
     npe03__Amount__c:
       random_number:
         min: 10
         max: 500
-    npsp__Day_of_Month__c:
-      random_number:
-        min: 1
-        max: 28
-    npsp__DisableFirstInstallment__c: true
-    npsp__StartDate__c: 
-      date_between:
-        start_date: 2020-01-01
-        end_date: today
-    npsp__InstallmentFrequency__c: 1
+
     npe03__Installment_Period__c: 
       random_choice:
         - Monthly
         - Yearly
         - Weekly
         - 1st and 15th
+    npsp__Day_of_Month__c:
+      if:
+        - choice:
+            when: ${{npe03__Installment_Period__c == 'Monthly'}}
+            pick: ${{random_number(min=1,max=29)}}
+        - choice:
+            when: ${{npe03__Installment_Period__c != 'Monthly'}}
+            pick: NULL
+
+    # uncomment this line if you want to disable the first installment of Opportunity Auto-Creation
+    #npsp__DisableFirstInstallment__c: true
+    npsp__StartDate__c: 
```

Comment:
@eehjunggnujhee  - this looks fantastic!! One minor comment -- what do you think about changing the npsp__StartDate__c to include some past dates as well. Since all of the dates being generated are from today and into the future, the Lapsed/Closed/Paused Statuses aren't always the most realistic. It would be really cool if those statuses were randomly assigned only when the date is before today. What do you think?

Reply:
Hello @jesslopez-sf 
Thank you for looking into it. That will be really nice if we can use probability for the random_choice.  I tried this logic before but did not take probability as valid syntax. Snowfakery picks the first choice when there is no probability. 

Like this:
```
- choice: 
            when: ${{date(npsp__StartDate__c) >= today }}
            pick: Active
            probability: 10%
```
Through, We can assign a value by a block of time like this as Paul recommended before:
```
        - choice: 
            when: ${{date(npsp__StartDate__c) <= today - relativedelta(months=6) }}
            pick: Closed
        - choice: 
            when: ${{date(npsp__StartDate__c) <= today - relativedelta(months=2) }}
            pick: Paused
```
The problem is .... I think the bug on random_date is not fixed on the month calculation. date_between works fine except 'm' calculation. It will output `today -1` for below. 
```
    npsp__StartDate__c: 
      date_between:
        start_date: -6m
        end_date: -1m (or today)
  ```
  `${{fake.date_between(start_date='-6m', end_date='today')}}`
This doesn't work as well. It became apparent that this might be the faker library error. So I will use either 'd' or 'y' for the date scope.
I will update the recipe today!

Reply:
Great explanation, @eehjunggnujhee and also a great solution. 🏆 🏆 🏆  Thank you!!!